File: Generator.php
<?php
require 'Parser.php';

class FactorioGenerator
{

    public $subgroups = [];
    public $groups = [];
    public $orders = [];
    public $machines = [];
    public $categories = [];
    public $recipes = [];
    public $resources = [];
    public $items = [];
    public $languages = [];
    public $translations = [];
    public $technologys = [];
    public $inserters = [];

    public $info = [];

    public $datastr = '';

    public $descPath = '.';
    public $basePath = '';
    public $firstSub = 0;

    public function __construct($descPath = '.')
    {
        $this->descPath = rtrim($descPath, '/');
    }

    public function parseMods($path, $data = false)
    {
        $path = rtrim($path, '/') . '/*';
        foreach (glob($path) as $mod) {
            $this->addMod($mod);
        }
        if ($data) {
            $this->firstSub = 1;
            foreach ($data as $types) {
                foreach ($types as $entity) {
                    $this->parseEntity($entity);
                }
            }
            foreach (glob($path) as $mod) {
                $this->parseMod($mod, true);
            }

        } else {
            $this->firstSub = 0;
            foreach (glob($path) as $mod) {
                $this->parseMod($mod);
            }
        }
        return $this;
    }

    public function addMod($path)
    {
        $info = json_decode(file_get_contents("{$path}/info.json"));
        $info->_path = $path;
        $this->info[$info->name] = $info;
    }

    public function parseMod($path, $only_locale = false)
    {
        if ($only_locale) {
            $dirs = ['locale'];
        } else {
            $dirs = ['locale', 'prototypes'];
        }
        foreach ($dirs as $dir) {
            $this->travel($path . '/' . $dir);
        }
    }

    public function save()
    {
        foreach ($this->subgroups as &$subgroup) {
            ksort($subgroup);
        }
        $this->writeJs('subgroups', $this->subgroups);
        $this->groups['technology']['icon'] = $this->groups['fluids']['icon'];
        $this->groups['technology']['order'] = 'y';
        $orders = [];
        foreach ($this->groups as $k => $group) {
            ksort($this->groups[$k]['subgroups']);
            $orders[] = $group['order'];
        }
        array_multisort($orders, $this->groups);

        $this->writeJs('groups', $this->groups);
        $this->writeJs('machines', $this->machines);

        foreach ($this->categories as $category => $content) {
            $orders = [];
            foreach ($content as $name) {
                if ($name == 'player') {
                    $orders[] = '0';
                } else {
                    $orders[] = $this->items[$name]['order'];
                }
            }
            array_multisort($orders, $this->categories[$category]);
        }
        $this->writeJs('categories', $this->categories);
        $this->writeJs('recipes', $this->recipes);
        $this->writeJs('resources', $this->resources);
        $this->writeJs('items', $this->items);
        $this->writeJs('technologys', $this->technologys);
        $this->writeJs('inserters', $this->inserters);
        foreach ($this->translations as $name => $content) {
            foreach ($content as $k => &$v) {
                if (is_array($v)) {
                    if (count(array_count_values($v)) == 1) {
                        $v = reset($v);
                    }
                }
            }
            $this->writeJs("translations/{$name}", $content, false, "translations['{$name}']=\n");
        }
        $this->languages = array_unique($this->languages);
        sort($this->languages);
        $this->writeJs('translations/list', $this->languages, true, "languages=\n");
        $this->writeData();
        if ($this->descPath != '.') {
            unlink($this->descPath . '/index.htm');
            unlink($this->descPath . '/index.js');
            unlink($this->descPath . '/index.css');
        }
        copy('index.htm', $this->descPath . '/index.htm');
        copy('index.js', $this->descPath . '/index.js');
        copy('index.css', $this->descPath . '/index.css');

    }

    public function convertPath($path, $is_source)
    {
        return preg_replace_callback('~__([^_]+?)__~', function ($matches) use ($is_source) {
            if (isset($this->info[$matches[1]])) {
                if ($is_source) {
                    return $this->info[$matches[1]]->_path;
                } else {
                    return $matches[1];
                }
            } else {
                return $matches[0];
            }
        }, $path);
    }

    public function saveIcon($file)
    {
        $converted = $this->convertPath($file, false);
        $source = $this->convertPath($file, true);
        $target = $this->descPath . '/graphics/' . $converted;
        if (!is_file($target)) {
            $dir = dirname($target);
            if (!is_dir($dir)) {
                mkdir($dir, 0755, true);
            }
            if (is_file($source)) {
                copy($source, $target);
            }
        }
        return 'graphics/' . $converted;
    }

    public function saveItem($entity)
    {
        if (!isset($entity['icon']) || !isset($entity['order'])) {
            return;
        }
        $item = [
            'order' => preg_replace('~\[.*?\]~', '', $entity['order']),
            'icon' => $this->saveIcon($entity['icon']),
        ];

        $this->items[$entity['name']] = $item;
    }

    public function saveResource($entity)
    {
        $resource = $this->buildItem($entity, [
            'category' => 'basic-solid',
        ]);
        $resource['hardness'] = $entity['minable']['hardness'];
        $resource['mining_time'] = $entity['minable']['mining_time'];
        $this->resources[$entity['name']] = $resource;

    }

    public function saveRecipe($entity)
    {
        $recipe = $this->buildItem($entity, [
            'result_count' => 1,
            'category' => 'crafting',
            'energy_required' => 0.5,
            'results',
        ]);
        $recipe['ingredients'] = [];
        if (isset($entity['ingredients'])) {
            $ingredients = $entity['ingredients'];
        } elseif (isset($entity['normal']['ingredients'])) {
            $ingredients = $entity['normal']['ingredients'];
        }
        // var_dump($entity);exit;
        foreach ($ingredients as $ingredient) {
            if (isset($ingredient['type'])) {
                $recipe['ingredients'][$ingredient['name']] = $ingredient['amount'];

            } else {
                $recipe['ingredients'][$ingredient[$this->firstSub]] = $ingredient[$this->firstSub + 1];
            }
        }
        $recipe['ingredient_count'] = count($recipe['ingredients']);
        if (isset($recipe['results']) && count($recipe['results']) == 1) {
            $recipe['result_count'] = $recipe['results'][$this->firstSub]['amount'];
            $recipe['type'] = $recipe['results'][$this->firstSub]['type'];
            unset($recipe['results']);
        }
        $this->recipes[$entity['name']] = $recipe;
    }

    public function saveMachine($entity)
    {

        $machine = $this->buildItem($entity, [
            'type',
            'crafting_speed' => 1,
            'mining_speed',
            'mining_power',
            'ingredient_count',
            'energy_source' => [],
            'energy_usage' => 0,
            'drain',
            'researching_speed',
        ]);
        if (
            isset($machine['energy_source']) &&
            isset($machine['energy_source']['emissions']) &&
            preg_match('~([0-9.]+)\s*/\s*([0-9.]+)~', $machine['energy_source']['emissions'], $match)
        ) {
            $machine['energy_source']['emissions'] = $match[1] / $match[2];
        }

        $machine['energy_usage'] = (int) $machine['energy_usage'];

        unset($machine['energy_source']['smoke']);
        unset($machine['energy_source']['fuel_inventory_size']);
        unset($machine['energy_source']['usage_priority']);
        isset($machine['energy_source']['type']) or $machine['energy_source']['type'] = 'biologic';
        isset($machine['energy_source']['emissions']) or $machine['energy_source']['emissions'] = 0;
        foreach (['resource_categories', 'crafting_categories'] as $category_group) {
            if (isset($entity[$category_group])) {
                foreach ($entity[$category_group] as $category) {
                    $this->categories[$category][] = $entity['name'];
                }
            }
        }

        $this->machines[$entity['name']] = $machine;

    }

    public function saveTechnology($entity)
    {
        $technology = $this->buildItem($entity, [
            'upgrade' => false,
        ]);
        $technology['prerequisites'] = [];
        if (isset($entity['prerequisites'])) {
            foreach ($entity['prerequisites'] as $prerequisite) {
                $technology['prerequisites'][] = $prerequisite;
            }

        }
        foreach ($entity['unit']['ingredients'] as $ingredient) {
            $technology['ingredients'][$ingredient[$this->firstSub]] = $ingredient[$this->firstSub + 1];

        }
        $technology['icon'] = $this->saveIcon($entity['icon']);
        $technology['order'] = preg_replace('~\[.*?\]~', '', $entity['order']);
        $firstOrder = preg_replace('~-.*~', '', $technology['order']);
        $technology['time'] = $entity['unit']['time'];
        $technology['count'] = 1; //$entity['unit']['count'];
        $technology['category'] = 'lab';
        $this->technologys[$entity['name']] = $technology;
        // $this->subgroups["technology-{$firstOrder}"][$entity['name']] = true;
        $this->groups['technology']['subgroups']["technology-{$firstOrder}"] = $firstOrder;

    }

    public function finishTechnologys()
    {
        $calcLevel = function ($name, &$technology) use (&$calcLevel) {
            $maxLevel = -1;
            foreach ($technology['prerequisites'] as $prerequisite_name) {
                $prerequisite = &$this->technologys[$prerequisite_name];
                if (!isset($prerequisite['level'])) {
                    $calcLevel($prerequisite_name, $prerequisite);
                }
                if ($prerequisite['level'] > $maxLevel) {
                    $maxLevel = $prerequisite['level'];
                }
            }
            if ($technology['upgrade']) {
                $technology['level'] = $maxLevel;
            } else {
                $technology['level'] = $maxLevel + 1;
            }

        };
        foreach ($this->technologys as $name => &$technology) {
            $this->subgroups['technology'][] = $name;
            $calcLevel($name, $technology);
            foreach ($technology['prerequisites'] as $prerequisite) {

            }
        }
    }

    public function saveInserter($entity)
    {
        $inserter = [];
        $angles = [];
        $distances = [];
        foreach (['pickup_position', 'insert_position'] as $position) {
            $x = $entity[$position][$this->firstSub];
            $y = $entity[$position][$this->firstSub + 1];

            if ($x == 0) {
                $angles[$position] = pi() / 2 * (($y > 0) - ($y < 0));
            } else {
                $angles[$position] = atan($y / $x);
            }

            $distances[$position] = sqrt($x * $x + $y * $y);
        }
        $angle = abs($angles['pickup_position'] - $angles['insert_position']);
        $rotation_turns_per_minute = 60/*seconds*/ * ($entity['rotation_speed'] * 60/* tick */) * pi() / $angle;

        $delta_distance = abs($distances['pickup_position'] - $distances['insert_position']);
        $extension_turns_per_minute = $delta_distance / $entity['extension_speed'] * 60;
        $inserter['turns_per_minute'] = min($rotation_turns_per_minute, $extension_turns_per_minute);

        // $inserter['icon'] = $this->saveIcon($entity['icon']);
        $this->inserters[$entity['name']] = $inserter;
    }

    public function saveLanguage($path)
    {

        $language = basename(dirname($path));
        $this->languages[] = $language;

        if (isset($this->translations[$language])) {
            $translation = $this->translations[$language];
        } else {
            $translation = [];

            if (is_file($ex = "locale/{$language}.ini")) {
                $translation = parse_ini_file($ex, false, INI_SCANNER_RAW);
            }

        }
        $locale = parse_ini_file($path, true, INI_SCANNER_RAW);
        foreach ([
            'item-name',
            'entity-name',
            'fluid-name',
            'equipment-name',
            'recipe-name',
            'technology-name',
        ] as $groupName) {
            if (isset($locale[$groupName])) {
                $group = $locale[$groupName];
                foreach ($group as $k => $v) {
                    $translation[$k][$groupName] = $v;
                }
            }
        }

        $this->translations[$language] = $translation;

    }

    public function parseEntity($entity)
    {
        $ksort_recursive = function (&$arr) use (&$ksort_recursive) {
            ksort($arr);
            foreach ($arr as $k => &$v) {
                if (is_array($v)) {
                    $ksort_recursive($v);
                }
            }
        };
        $ksort_recursive($entity);

        $checkGroup = function ($entity) {
        };
        if (!isset($entity['type'])) {
            return;
        }
        switch ($entity['type']) {
            case 'item-group':
            case 'item-subgroup':
                if ($entity['type'] == 'item-group') {
                    $groupName = $entity['name'];
                } else {
                    $groupName = $entity['group'];
                }
                if (!isset($this->groups[$groupName])) {
                    $this->groups[$groupName] = ['icon' => '', 'order' => '', 'subgroups' => []];
                }
                if ($entity['type'] == 'item-group') {

                    $this->groups[$entity['name']]['icon'] = $this->saveIcon($entity['icon']);
                    $this->groups[$entity['name']]['order'] = $entity['order'];
                } else {

                    $this->groups[$entity['group']]['subgroups'][$entity['name']] = $entity['order'];
                }
                break;
            case 'inserter':
                $this->saveInserter($entity);
                break;
            case 'technology':
                $this->saveTechnology($entity);
                break;
            case 'lab':
                $this->categories['lab'][] = $entity['name'];
                $this->saveMachine($entity);
                break;
            case 'resource':
            case 'recipe':
                $this->{"save{$entity['type']}"}($entity);
            default:
                $this->saveItem($entity);
                break;

        }
        if (isset($entity['subgroup']) && !isset($this->subgroups[$entity['subgroup']][$entity['name']])) {
            $this->subgroups[$entity['subgroup']][$entity['name']] = true;
        }

        if (isset($entity['crafting_categories']) || isset($entity['resource_categories'])) {
            $this->saveMachine($entity);
        }

    }
    public function travel($currentPath)
    {
        foreach (glob($currentPath) as $path) {
            if (is_dir($path)) {
                $this->travel($path . '/*');
                continue;
            }
            if ('cfg' == $ext = pathinfo($path, PATHINFO_EXTENSION)) {
                $this->saveLanguage($path);
            } elseif ('lua' == $ext) {
                echo "Parsing $path...\n";
                $result = ((new FactorioLuaParser)->parseFile($path));
                if (!$result) {
                    continue;
                }

                foreach ($result as $field) {
                    if (!isset($field[0]) || !isset($field[0][0])) {
                        continue;
                    }

                    foreach ($field[0][0] as $entity) {
                        $this->parseEntity($entity);
                    }
                }
            }
        }
    }

    protected function buildItem(&$source, $params)
    {
        if (isset($source['normal'])) {
            foreach ($source['normal'] as $k => $v) {
                $source[$k] = $v;
            }
        }
        $target = [];
        foreach ($params as $param => $default) {
            $has_default = true;
            if (is_int($param)) {
                $param = $default;
                $has_default = false;
            }
            if (isset($source[$param])) {
                $target[$param] = $source[$param];
            } else {
                $has_default and $target[$param] = $default;
            }
        }
        return $target;
    }

    protected function writeJs($name, $content, $concat = true, $prefix = '', $postfix = '')
    {
        if ($sort) {
            ksort($content);
        }

        if ($prefix === '') {
            $prefix = "$name=\n";
        }

        if ($concat) {
            $this->datastr .= $prefix . json_encode($content, JSON_UNESCAPED_UNICODE) . "\n";
        }
        if (!is_dir(dirname("{$this->descPath}/{$name}"))) {
            mkdir(dirname("{$this->descPath}/{$name}"), 0755, true);
        }
        unlink("{$this->descPath}/{$name}.js");
        file_put_contents("{$this->descPath}/{$name}.js", $prefix . json_encode($content, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT) . "\n") . $postfix;
        var_dump($name);
    }

    protected function writeData()
    {
        unlink("{$this->descPath}/data.js");
        file_put_contents("{$this->descPath}/data.js", $this->datastr);
    }
}

File: core/generator.lua
local fs = require 'fs'
local dkjson = require 'dkjson'
local generator = {}
local pathMapping = {}

local legalTypes = {
	'item',
	'armor',
	'ammo',
	'capsule',
	'fluid',
	'gun',
	'item-group',
	'item-with-entity-data',
	'mining-tool',
	'module',
	'player',
	'rail-planner',
	'recipe',
	'repair-tool',
	'resource',
	'technology',
	'tool',
}
local typesLength = 17

local meta

local icons = {}
local icons_to_copy = {}

local function init()
	icons = {}
	icons_to_copy = {}
	pathMapping = {}

	meta = {
		groups = {},
		subgroups = {},
		items = {},
		inserters = {},
		recipes = {
			dummy = {
				name = 'dummy',
				category = 'dummy',
				normal = {
					energy_required = 0.5,
					results = {},
					ingredients = {},
					ingredient_count = 0,
				}
			},
		},
		modules = {},
		beacons = {},
		resources = {},
		categories = {
			dummy = {'dummy'},
		},
		machines = {
			{
				name = 'dummy',
				type = 'dummy',
				module_slots = 0,
				crafting_speed = 0,
				energy_usage = 0,
				allowed_effects = {},
				energy_source = {
					type = 'biologic'
				},
			}
		},
		technologies = {},
		languages = {},
		translations = {},
	}
end

local function buildItem(source, params)
	target = {}
	target.name = source.name
	for param, default in pairs(params) do
		local has_default = true
		if is_int(param) then
			param = default
			has_default = false
		end
		if not (source[param] == nil) then
			target[param] = source[param]
		else
			if has_default then target[param] = default end
		end
	end
	return target
end

local function copyIcon(entity, name)
	if name then
		icons_to_copy[name] = true
	end
	if entity then
		icons_to_copy[entity.name] = true
	end
end

local function translateIconPath(entity)
	local icon = entity.icon
	if (entity.icons) then
		icon = entity.icons[1].icon
	end
	if not icon then return end
	return icon:gsub('^__%w+__', pathMapping)
end

local function saveItem(entity)
	local path = translateIconPath(entity)

	icons[entity.name] = icons[entity.name] or {}
	icons[entity.name][entity.type] = path

	local name = entity.name

	if entity.minable and entity.minable.result then
		name = entity.minable.result
	end

	local item = {}
	if meta.items[name] then item = meta.items[name] end

	if not item.icon then item.icon = path end
	if not item.name then item.name = name end
	if not item.type then item.type = entity.type end
	if not item.order then item.order = entity.order end
	meta.items[name] = item
	return path
end

local function finalize()

	local files = {}
	for name in pairs(icons_to_copy) do
		local path
		for i = 1, typesLength do
			if not icons[name] then break end
			path = icons[name][legalTypes[i]]
			if path then
				break
			end
		end
		if not path then
			goto continue
		end
		if not fs.exists(path) and isNode() then
			print(name, path)
			error()
		end

		files[path] = true

		::continue::
	end
	return files
end

local function saveLanguages(moduleName)
	print('loading language ' .. moduleName)
	-- Will be called when parsing a mod
	-- The working directory is different
	for _, language in pairs(fs.readDir('data/' .. moduleName .. '/locale')) do
		for _, filename in pairs(fs.readDir('data/' .. moduleName .. '/locale/' .. language)) do
			local fullname = 'data/' .. moduleName .. '/locale/' .. language .. '/' .. filename
			if filename == 'info.json' then
				zipIt(fullname)
				local content = fs.readFile(fullname)
				local name = dkjson.decode(content)['language-name']
				meta.languages[language] = name
			elseif filename:sub(-4) == '.cfg' then
				zipIt(fullname)
				local ini = loadINI(fullname)
				meta.translations[language] = meta.translations[language] or {}
				for _, groupName in pairs({
						'gui-menu',
						'gui-technology',
						'item-name',
						'entity-name',
						'fluid-name',
						'equipment-name',
						'recipe-name',
						'technology-name',
						''
						}) do
					if ini[groupName] then
						if not meta.translations[language][groupName] then
							meta.translations[language][groupName] = {}
						end
						local group = ini[groupName]
						for k, v in pairs(group) do
							meta.translations[language][groupName][k] = v
						end
					end
				end
			end
		end
	end
end

local function saveQuantorioLanguages()
	for _, file in pairs(fs.readDir('locale')) do
		local language = file:match('(.-)%.')
		if meta.translations[language] then
			local ini = loadINI('locale/' .. file)
			for _, group in pairs(ini) do
				for k, v in pairs(group) do
					meta.translations[language][k] = v
				end
			end
		end
	end
end

local function saveGroup(entity)
	if not meta.groups[entity.name] then
		meta.groups[entity.name] = {
			name = entity.name,
			subgroups = {},
		}
	end
	meta.groups[entity.name].icon = translateIconPath(entity)
	copyIcon(entity)
end

local function saveSubGroup(entity)
	meta.groups[entity.group].subgroups[entity.name] = meta.groups[entity.group].subgroups[entity.name] or {}
	meta.subgroups[entity.name] = entity
end

local function saveInserter(entity)
	local inserter = {}
	local angles = {}
	local distances = {}
	local x, y
	for _, position in pairs({'pickup_position', 'insert_position'}) do
		x = entity[position][1]
		y = entity[position][2]

		if (x == 0) then
			local sign = 0
			if (y > 0) then
				sign = 1
			end
			if (y < 0) then
				sign = -1
			end
			angles[position] = math.pi / 2 * sign
		else
			angles[position] = math.atan(y / x)
		end

		distances[position] = math.sqrt(x * x + y * y)
	end
	angle = math.abs(angles.pickup_position - angles.insert_position)
	rotation_turns_per_minute = 60 --[[seconds]] * (entity.rotation_speed * 60 --[[tick]]) * math.pi / angle

	delta_distance = math.abs(distances.pickup_position - distances.insert_position)
	extension_turns_per_minute = delta_distance / entity.extension_speed * 60
	inserter.turns_per_minute = math.min(rotation_turns_per_minute, extension_turns_per_minute)
	inserter.name = entity.name
	inserter.stack = entity.stack or false

	table.insert(meta.inserters, inserter)
end

local function saveModule(entity)
	local module = buildItem(entity, {
		'type',
		'effect',
		'order',
		'limitation',
	})
	table.insert(meta.modules, module)

end

local function saveBeacon(entity)
	local beacon = buildItem(entity, {
		'type',
		'energy_usage',
		'distribution_effectivity',
	})
	beacon.allowed_effects = {}
	for _, effect in pairs(entity.allowed_effects) do
		table.insert(beacon.allowed_effects, effect)
	end
	beacon.module_slots = entity.module_specification.module_slots
	table.insert(meta.beacons, beacon)
end

local function saveResource(entity)
	local resource = buildItem(entity, {
		category = 'basic-solid',
	})
	resource.hardness = entity.minable.hardness
	resource.mining_time = entity.minable.mining_time
	if entity.minable.required_fluid then
		resource.fluid_amount = entity.minable.fluid_amount
		resource.required_fluid = entity.minable.required_fluid
	end
	meta.resources[entity.name] = resource
	copyIcon(entity)
end

local function saveRecipe(entity)
	if entity.subgroup == 'empty-barrel' or entity.subgroup == 'fill-barrel' then return end
	local recipe = buildItem(entity, {
		category = 'crafting',
	})

	local all = {
		normal = entity
	}

	if entity.normal then
		all.normal = entity.normal
	end
	if entity.expensive then
		all.expensive = entity.expensive
	end

	local lastResult = nil
	local resultTypeCount = 0

	for difficulty, config in pairs(all) do
		recipe[difficulty] = {
			ingredients = {},
		}
		local count = 0
		for _, ingredient in pairs(config.ingredients) do
			count = count + 1
			if ingredient.type then
				recipe[difficulty].ingredients[ingredient.name] = ingredient.amount
				copyIcon(ingredient)
			else
				recipe[difficulty].ingredients[ingredient[1]] = ingredient[2]
				copyIcon(nil, ingredient[1])
			end
		end
		recipe[difficulty].ingredient_count = count
		local results = {}
		if config.result then
			results[config.result] = config.result_count or 1
			lastResult = config.result
			resultTypeCount = 1
		else
			for _, result in pairs(config.results) do
				results[result.name] = result.amount
				lastResult = result.name
				resultTypeCount = resultTypeCount + 1
			end
		end
		results.type = nil
		recipe[difficulty].results = results
		recipe[difficulty].energy_required = config.energy_required or 0.5
	end

	local name
	if lastResult and lastResult ~= name and resultTypeCount == 1 then
		name = lastResult
		recipe.showName = name
	end

	meta.recipes[entity.name] = recipe

	copyIcon(entity, name)
end

local function saveMachine(entity)
	if entity.name == 'default' and entity.type == 'god-controller' then return end

	local machine = buildItem(entity, {
		'type',
		crafting_speed = 1,
		'mining_speed',
		'mining_power',
		'ingredient_count',
		energy_source = {},
		energy_usage = 0,
		'drain',
		'researching_speed',
	})
	if entity.module_specification and entity.module_specification.module_slots then
		machine.module_slots = entity.module_specification.module_slots
	else
		machine.module_slots = 0
	end

	if entity.pumping_speed then
		entity.resource_categories = {entity.fluid}
		machine.crafting_speed = entity.pumping_speed
		if not meta.recipes[entity.fluid] then
			local recipe =
				{
					name = entity.fluid,
					category = entity.fluid,
					normal = {
						energy_required = 1,
						results = {},
						ingredients = {},
						ingredient_count = 0,
					}
				}
			recipe.normal.results[entity.fluid] = 60
			recipe.showName = entity.fluid
			meta.recipes[entity.fluid] = recipe
		end
	end

	machine.energy_usage = tonumber(string.sub(machine.energy_usage, 0, -3)) or 0

	machine.energy_source.smoke = nil
	machine.energy_source.fuel_inventory_size = nil
	machine.energy_source.usage_priority = nil
	machine.energy_source.type = machine.energy_source.type or 'biologic'
	machine.energy_source.emissions = machine.energy_source.emissions or 0

	for _, category_group in pairs({'resource_categories', 'crafting_categories'}) do
		if entity[category_group] then
			for _, category in pairs(entity[category_group]) do
				if not meta.categories[category] then
					meta.categories[category] = {}
				end
				table.insert(meta.categories[category], entity.name)
			end
		end
	end
	machine.allowed_effects = {}
	if entity.allowed_effects then
		for _, effect in pairs(entity.allowed_effects) do
			table.insert(machine.allowed_effects, effect)
		end
	end
	table.insert(meta.machines, machine)
	copyIcon(entity)
end

local function saveTechnology(entity)
	local technology = buildItem(entity, {
		'type',
		'effects',
		'order',
		'prerequisites',
	})
	meta.technologies[entity.name] = technology
end

local function pushGroup(entity)
	if not entity.subgroup then return end
	meta.groups[meta.subgroups[entity.subgroup].group].subgroups[entity.subgroup][entity.name] = true
end

local function parse(data, m)
	-- global
	pathMapping = m

	local mapping = {
		{ name = 'inserter', saver = saveInserter },
		{ name = 'module', saver = saveModule },
		{ name = 'beacon', saver = saveBeacon },
		{ name = 'resource', saver = saveResource },
		{ name = 'recipe', saver = saveRecipe },
		{ name = 'fluid', saver = copyIcon },
		{ name = 'player', saver = saveMachine },
		{ name = 'assembling-machine', saver = saveMachine },
		{ name = 'furnace', saver = saveMachine },
		{ name = 'mining-drill', saver = saveMachine },
		{ name = 'rocket-silo', saver = saveMachine },
		{ name = 'offshore-pump', saver = saveMachine },
		{ name = 'technology', saver = saveTechnology },
	}

	local mappingLength = 13

	for _, entity in pairs(data['item-group']) do
		saveItem(entity)
		saveGroup(entity)
	end

	for _, entity in pairs(data['item-subgroup']) do
		saveItem(entity)
		saveSubGroup(entity)
	end

	for i = 1, typesLength do
		local typeName = legalTypes[i]
		local entities = data[typeName]
		for _, entity in pairs(entities) do
			saveItem(entity)
			pushGroup(entity)
		end
	end

	for i = 1, mappingLength do
		local typeName = mapping[i].name
		local saver = mapping[i].saver
		local entities = data[typeName]

		for _, entity in pairs(entities) do
			if saver then
				saver(entity)
			end
			pushGroup(entity)
		end
	end
end

return {
	init = init,
	parse = parse,
	finalize = finalize,
	saveLanguages = saveLanguages,
	saveQuantorioLanguages = saveQuantorioLanguages,
	getMeta = function ()
		return meta
	end,
}
